Add command line option --stdin to read from standard input.

miguelbarao opened the pull request:
- If yaml files are given as arguments, parses the files first and then stdin.
- If no yaml files are given, just checks stdin.
- If yamllint is run without any files or --stdin option, just finishes without errors.

I found the option --stdin to be more reasonable than with no args. Mainly when automating yamllint from scripts.

Tried to minimize changes to base code. But had to create a function `show_problems` which is called for files and for stdin with slightly different arguments. This function is basically copy/paste of existing code.

Also moved a lot of code from the try/exception to the else block, so that it only catches on open(file). You might want to check on that (are there other reasons for an EnvironmentError exception?).

adrienverge commented:
Thanks for contributing!

> If no yaml files are given, just checks stdin.

From what I read from your code, this seems wrong. Do we agree that `--stdin` is required to lint from standard input?

Can you add tests to `tests/test_cli.py` to make sure **all** possible cases are tested? In particular, passing `--stdin` + files should fail with an explicit error.

miguelbarao commented:
Thanks for your quick reply.

Yes, —stdin is required to read from stdin.

I’ll try to write the tests in the next few days.



> On 16 Jan 2019, at 18:48, Adrien Vergé <notifications@github.com> wrote:
> 
> Thanks for contributing!
> 
> If no yaml files are given, just checks stdin.
> 
> From what I read from your code, this seems wrong. Do we agree that --stdin is required to lint from standard input?
> 
> Can you add tests to tests/test_cli.py to make sure all possible cases are tested? In particular, passing --stdin + files should fail with an explicit error.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/adrienverge/yamllint/pull/160#issuecomment-454894743>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AF0m8k1fPvITse3XkKLg_28OHIlDutMzks5vD3QDgaJpZM4aDsFB>.
>

momoneko commented:
@miguelbarao, are you still working on this? Need this feature too for a project.

Just a note: the unix philosophy generally dictates that when a cli utility is started without args or with `-` as an argument to read from stdin (see cat or grep). Not terribly important, but something to keep in mind.

adrienverge commented:
Good point @momoneko, `-` instead of `--stdin` would be Unix-compliant and elegant :+1:

miguelbarao commented:
Agree!

Added single dash option so that both `yamllint -` and `yamllint --stdin` read from standard input (I kept `--stdin` because most commands have both a short and long version, but can be easily removed if you want).

I also added a test in `test_cli.py` to check if the standard input is being processed correctly.
It sends an invalid yaml string to stdin and checks if yamllint reports the correct error.

@adrienverge, you referred that reading from stdin and files should be an error. Why? Everything is a file and works the same. We can do something like

```sh
cat filelist.txt | xargs yamllint - another.yaml
```

to check all files from filelist.txt and another.yaml.
I didn't prevent this usage.

adrienverge commented:
@miguelbarao thanks!

Yes I prefer to remove `--stdin` for simplicity and consistency with Unix tools.

> you referred that reading from stdin and files should be an error. Why?

My concern is that the filename for standard input is not defined, hence, it can produce ambiguities in the output. For example running:
```
$ cat standard-input.yaml | xargs yamllint - stdin
```
... can produce:
```
stdin
  5:6       warning  missing starting space in comment  (comments)
stdin
  1:1       warning  missing document start "---"  (document-start)
```

Solutions I see:
- disable the use of `-` + files for now (while leaving the door open to enable it later -- there won't be legacy problems)
- enable it, but replace `stdin` by `-` in the CLI output.

What do you think?

miguelbarao commented:
I wouldn't use a single `-` for several reasons: it can be confused with the underscore, it will appear underlined, and there can be files named with a single dash that yamllint can already process with `yamllint -- -`.

To prevent ambiguity, the only way I see is the stdin name to include invalid chars for a filename.
The only forbidden visible char is the slash `/`, so writing `/stdin/` should be safe.

adrienverge commented:
Good point for `yamllint -- -`.

In this case, I prefer that we prevent the usage of `-` + files, for now:
- it will still be feasible in the future without breaking usage,
- it will allow us to merge this PR soon, while we collect other feedbacks and find a solution everybody agrees on,
- the usage of `-` + files must be very rare in real-life use cases.

miguelbarao commented:
Ok, I think it's now running as intended:

- dash option `-` to read from stdin (removed `--stdin`).
- stdin and files are mutually exclusive.
- if no `-` and no files are specified, terminates with an error and usage message like in the original version.

I also added code to the `test_run_with_bad_arguments` function to check the mutually exclusive case.
The other cases are already covered.

Thanks for your support.

miguelbarao commented on yamllint/cli.py:
```diff
@@ -138,32 +166,21 @@ def run(argv=None):
         except EnvironmentError as e:
             print(e, file=sys.stderr)
             sys.exit(-1)
+        else:
```
Not strictly required, but I think it's good practice, since changes in the `except` clauses do not affect the `else`.

adrienverge commented:
(Please excuse me for the delay, I'm a bit busy currently.)

Thanks a lot for the good work: the PR is neat and this is a cool feature!

My only concern is that OS errors (interruptions, broken pipes, etc.) aren't caught anymore while reading the file (only when opening it, and only for real files, not standard input).
To mitigate that, I allowed myself to push to your branch with a proposal (commit `[PROPOSAL] Keep catching interruptions, broken pipes and other OS errors`).

If you're OK with it, you can squash all commits (or I'll do it) and then it will be good for a new release :)

adrienverge commented:
Alright!

Thanks for your contribution to yamllint :+1:
